build/teamcity/cockroach/nightlies/cloud_unit_tests_impl.sh

```diff
@@ -25,7 +24,7 @@ export AWS_CONFIG_FILE="$PWD/.aws/config"
 log_into_aws
 
 exit_status=0
-$BAZEL_BIN/pkg/cmd/bazci/bazci_/bazci -- --config=ci \
+$BAZEL_BIN/pkg/cmd/bazci/bazci_/bazci --go_test_json_output_file=$GO_TEST_JSON_OUTPUT_FILE -- --config=ci \
```

Comment:
So IMO it's pretty odd that the callers would specify `go_test_json_output_file` when they don't actually care about the location of the file, or even that it exists. I think we should do one of the following:

1. Instead, make the command-line flag something like `--github-post`. In this case `bazci` should automatically choose an appropriate location for the `GO_TEST_JSON_OUTPUT_FILE` and process it "under the hood" without requiring the caller to have any knowledge of the process at all. The change is valuable by itself, but also we want to change `bazci` so that it processes these reports based on what's in the `test.xml`, NOT what's in the `GO_TEST_JSON_OUTPUT_FILE`. (`GO_TEST_JSON_OUTPUT_FILE` is a hack that ultimately we would want to get rid of.) So doing this allows us to make that change without having to update any of the callers, which is ideal.
2. ... or, instead of doing that, just have `bazci` "know" when it needs to post issues to GitHub and silently make the "right" decision instead of requiring the caller specify that GitHub issues should be posted. I *think* it is sufficient to post GitHub issues iff the `TC_BUILD_BRANCH` is `master`, or maybe just if `GITHUB_API_TOKEN` is set? This would mean that failed unit tests in Bazel Essential CI start posting GitHub issues, which I think is the desired behavior(?) @srosenberg may have thoughts about how this should work.

Reply:
> post GitHub issues iff the TC_BUILD_BRANCH is master, release-*, or provisional

I think this is true but other steps such as creating `failures.txt` file and creating the failures tarball do not check which branch we are on so I am leaning towards option "1". Should we use `--process-tests-failures` instead of `--github-post` though? Posting github issues is only one of the things this function does.

Reply:
ok so just noticed a use-case where it's useful to have pass the output file name to `bazci`:

https://github.com/cockroachdb/cockroach/pull/87938/files#diff-aae2b85e528bc2495839df3b394be98b3190bcbcdf0fb4f0918ab61da73bb2ff

Reply:
> I think this is true but other steps such as creating failures.txt file and creating the failures tarball do not check which branch we are on so I am leaning towards option "1".

I'm wondering why it matters. Is there any reason NOT to create the `failures.txt` file or the failures tarball? We can just always unconditionally do that, and then only do the "posting to GitHub" part if we're able (if we have the necessary tokens and are on a correct branch).

I'm also wondering if the `failures.txt` file and the failures tarball are really any use to us at all post-Bazel, but maybe @srosenberg has some insight into what purpose they're meant to serve.

> ok so just noticed a use-case where it's useful to pass the output file name to bazci:

Hmm. I don't think so. The different filenames are just there because we're running the test more than once and we want a different underlying file per test (so we don't accidentally report the same failed test twice). But `bazci` might as well just do it under the hood as mentioned.

Reply:
> Hmm. I don't think so. The different filenames are just there because we're running the test more than once and we want a different underlying file per test (so we don't accidentally report the same failed test twice). But bazci might as well just do it under the hood as mentioned.

So we just need to add a random suffix after some default path and default basename like `artifactsdir/go_test_json_default_name.randomsuffix`? My point here is "we don't care if the name is meaningful/still refers to the test"?

> I'm also wondering if the failures.txt file and the failures tarball are really any use to us at all post-Bazel, but maybe @srosenberg has some insight into what purpose they're meant to serve.

Do we still run `make` nightlies? If so, what you mentioned is a future action right? 

Here are the changes I will do:

1. always create `failures.txt` if the command is `test`
2. always create tarball if the command is `test` and it returns an error
3. post github issues if the command is test and the branch is in `master, release-*, or provisional`
4. no flags needed

How does that sound?

Reply:
Actually, I've just realized you're right, we can't do the automatic issue processing until we have moved off `GO_TEST_JSON_OUTPUT_FILE`. The problem is that `bazel` can run tests in parallel and if multiple tests are writing in parallel to the same `GO_TEST_JSON_OUTPUT_FILE` then we'll get unparseable garbage. So guarding this behavior behind a flag like `--process-test-failures` is actually preferable for now. Here is what I am thinking in terms of future work:

1. In this PR, add a `--process-test-failures` flag which is called by nightlies. When this flag is set, create failures.txt/the failures tarball and post GitHub issues if appropriate (similar behavior to option (1) from my original comment).
2. Follow-up: move `bazci` to streaming rather than processing all `test.xml`'s after all tests have finished.
3. Teach `bazci` how to process test failures from `test.xml` instead of Go `test.json` files.
4. Remove the `--process-test-failures` flag and have `bazci` infer which post-processing actions need to be taken.

Reply:
> In this PR, add a --process-test-failures flag which is called by nightlies. When this flag is set, create failures.txt/the failures tarball and post GitHub issues if appropriate (similar behavior to option (1) from my original comment).

In this case should the json file name be generated by bazci by adding the following suffix to some default basename:
`strconv.Itoa(rand.New(rand.NewSource(time.Now().UnixNano())).Intn(1000000)))`
Any better ways?

> Follow-up: move bazci to streaming rather than processing all test.xml's after all tests have finished.

I can work on this next week but I remember you mentioned it before so just making sure you are not currently working on it right?

Reply:
> Any better ways?

This looks fine, but consider just using `os.CreateTemp()` for this.

> I can work on this next week but I remember you mentioned it before so just making sure you are not currently working on it right?

I have a draft PR but I probably won't finish working on it this week, I can send you what I have so far.

Reply:
> consider just using os.CreateTemp() for this.

nice, more reliable because it ensures the file doesn't exist